daemon/emer-daemon.c

```diff
@@ -1665,3 +2072,67 @@ emer_daemon_get_permissions_provider (EmerDaemon *self)
   EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
   return priv->permissions_provider;
 }
+
+void
+emer_daemon_start_aggregate_timer (EmerDaemon         *self,
+                                   EmerAggregateTimer *timer,
+                                   uid_t               unix_user_id,
+                                   GVariant           *event_id,
+                                   GVariant           *aggregate_key,
+                                   gboolean            has_payload,
+                                   GVariant           *payload)
+{
+  EmerDaemonPrivate *priv;
+  g_autoptr(GVariant) cache_key = NULL;
+  AggregateTimerData *timer_data;
+
+  g_return_if_fail (EMER_IS_DAEMON (self));
+  g_return_if_fail (event_id != NULL);
+  g_return_if_fail (aggregate_key != NULL);
+
+  priv = emer_daemon_get_instance_private (self);
+
+  if (!is_uuid (event_id))
+    {
+      g_warning ("Event ID must be a UUID represented as an array of %"
+                 G_GSIZE_FORMAT " bytes. Dropping event.", UUID_LENGTH);
+      return;
+    }
+
+  cache_key = g_variant_new ("(u@ayvmv)",
+                             unix_user_id,
+                             event_id,
+                             aggregate_key,
+                             get_nullable_payload (payload, has_payload));
+
+  if (g_hash_table_contains (priv->aggregate_timers, cache_key))
+    {
+      g_warning ("There already is a timer for this event and user");
+      return;
+    }
+
+  g_variant_ref_sink (cache_key);
+
+  timer_data = g_new0 (AggregateTimerData, 1);
+  timer_data->daemon = self;
+  timer_data->start_monotonic_us = g_get_monotonic_time ();
+  timer_data->unix_user_id = unix_user_id;
+  timer_data->cache_key = g_variant_ref (cache_key);
+  timer_data->aggregate_key = g_variant_ref_sink (aggregate_key);
+  timer_data->event_id = g_variant_ref_sink (event_id);
+  timer_data->payload = get_nullable_payload (payload, has_payload);
+
+  g_signal_connect (timer,
+                    "handle-stop-timer",
+                    G_CALLBACK (on_timer_stopped_cb),
+                    timer_data);
+
+  g_hash_table_insert (priv->aggregate_timers,
+                       g_variant_ref (cache_key),
+                       timer_data);
+
+  if (priv->dispatch_aggregate_timers_daily_source_id == 0)
+    schedule_next_midnight_tick (self);
+  if (priv->dispatch_aggregate_timers_monthly_source_id == 0)
+    schedule_next_month_tick (self);
```

Comment:
So here we are only scheduling these ticks if at least 1 timer is running. I think it might be clearer to make them unconditional while the daemon is running. On start-up it needs to check for missed ticks anyway.